docs/ONNX_Runtime_Graph_Optimizations.md

```diff
@@ -0,0 +1,122 @@
+# Graph Optimizations in ONNX Runtime
+
+ONNX Runtime provides various graph level optimizations to improve model performance. Graph optimizations are essentially graph level rewrites ranging from small graph simplifications and node eliminations to more complex node fusions and layout optimizations. 
+
+Graph optimizations are divided in several levels\categories based on their complexity and function. They can be performed both online and offline. ONNX Runtime provides Python, C#, C++ and C apis to enable different optimization levels and to choose the offline vs online mode. In offline mode the runtime saves the optimized graph to the disk. **Note**: When running in offline mode make sure to use the exact same options and hardware as the target machine.
+
+## Graph Optimization Levels
+
+Graph optimizations are divided in 3 levels
+* Basic
+* Extended
+* Layout Optimizations
+
+### Basic Graph Optimizations
+
+These are semantics preserving graph rewrites which remove redundant nodes and redundant computations. These optimizations are enabled by default. They run before graph partitioning and thus apply to all the execution providers. Available basic graph optimizations are as follows:
+
+* Constant Folding : Statically computes parts of the graph that rely only on constant initializers. This eliminates the need to compute them during runtime.
+
+* Redundant node eliminations : Removes all redeundant nodes without changing the graph structure
+  * Identity Elimination
+  * Slice Elimination
+  * Unsqueeze Elimination
+  * Dropout Elimination
+
+* Semantics preserving node fusions : Fuses\folds multiple nodes into 1 node. For example Conv Add fusion folds the Add operator as the bias of the Conv operator.
+  * Conv Add Fusion
+  * Conv Mul Fusion
+  * Conv BatchNorm Fusion
+  * Relu Clip Fusion
+
+### Extended Graph Optimizations
+
+These optimizations include complex node fusions. They are run after graph partitioning and are only applied to the nodes assigned to CPU execution provider. Available extended graph optimizations are as follows:
+
+* GEMM Activation Fusion
+* Matmul Add Fusion
+* Conv Activation Fusion
+* GELU Fusion
+
+### Layout Optimizations
+
+These optimizations change the data layout for applicable nodes to achieve higher performance improvements. They are run after graph partitioning and are only applied to nodes assigned to CPU execution provider. Available layout optimizations are as follows:
+
+* NCHWc Optimizer : Optimizes the graph by using NCHWc layout instead of NCHW layout.
+
+## Usage
+
+### General Note
+Levels: 
+ONNX Runtime defines GraphOptimizationLevel enum for these levels. Each level when enabled, enables all the preceeding levels too. Mapping of these levels to enum is as follows:
```

Comment:
Consider rewording: "Enabling each level also enables all preceding levels. For example, enabling level 3 also enables levels 2 and 1."